elytron/src/main/java/org/wildfly/extension/elytron/FileSystemRealmDefinition.java

```diff
@@ -96,21 +115,62 @@
+                    .setValidator(new StringAllowedValuesValidator(BASE64, HEX))
+                    .setAllowExpression(true)
+                    .setFlags(AttributeAccess.Flag.RESTART_RESOURCE_SERVICES)
+                    .build();
+
+    static final SimpleAttributeDefinition HASH_CHARSET =
+            new SimpleAttributeDefinitionBuilder(ElytronDescriptionConstants.HASH_CHARSET, ModelType.STRING, true)
+                    .setFlags(AttributeAccess.Flag.RESTART_RESOURCE_SERVICES)
+                    .setDefaultValue(new ModelNode(UTF_8))
+                    .setValidator(new CharsetValidator())
+                    .setAllowExpression(true)
+                    .build();
+
+    static final SimpleAttributeDefinition CREDENTIAL_STORE =
+            new SimpleAttributeDefinitionBuilder(ElytronDescriptionConstants.CREDENTIAL_STORE, ModelType.STRING, false)
+                    .setAllowExpression(true)
+                    .setRequired(false)
+                    .setRequires(ElytronDescriptionConstants.SECRET_KEY)
+                    .setMinSize(1)
+                    .setRestartAllServices()
+                    .setCapabilityReference(CREDENTIAL_STORE_CAPABILITY, SECURITY_REALM_CAPABILITY)
+                    .build();
 
-    static final SimpleAttributeDefinition HASH_CHARSET = new SimpleAttributeDefinitionBuilder(ElytronDescriptionConstants.HASH_CHARSET, ModelType.STRING, true)
-            .setFlags(AttributeAccess.Flag.RESTART_RESOURCE_SERVICES)
-            .setDefaultValue(new ModelNode(UTF_8))
-            .setValidator(new CharsetValidator())
-            .setAllowExpression(true)
-            .build();
+    static final SimpleAttributeDefinition SECRET_KEY =
+            new SimpleAttributeDefinitionBuilder(ElytronDescriptionConstants.SECRET_KEY, ModelType.STRING, false)
+                    .setAllowExpression(true)
+                    .setRequired(false)
+                    .setRequires(ElytronDescriptionConstants.CREDENTIAL_STORE)
+                    .setMinSize(1)
+                    .setRestartAllServices()
+                    .build();
+
+    static final SimpleAttributeDefinition KEY_STORE =
+            new SimpleAttributeDefinitionBuilder(ElytronDescriptionConstants.KEY_STORE, ModelType.STRING, false)
+                    .setAllowExpression(true)
+                    .setRequired(false)
+                    .setRequires(ElytronDescriptionConstants.KEY_STORE_ALIAS)
+                    .setMinSize(1)
+                    .setRestartAllServices()
+                    .setCapabilityReference(KEY_STORE_CAPABILITY, SECURITY_REALM_CAPABILITY, true)
+                    .setFlags(AttributeAccess.Flag.RESTART_RESOURCE_SERVICES)
+                    .build();
+
+    static final SimpleAttributeDefinition KEY_STORE_ALIAS =
+            new SimpleAttributeDefinitionBuilder(ElytronDescriptionConstants.KEY_STORE_ALIAS, ModelType.STRING, false)
+                    .setAllowExpression(true)
```

Comment:
Since `key-store` and `key-store-alias` both allow expressions, this should also be tested in `realms-test.xml`.

Reply:
Added